File: level_1/turtlesim/ros2_ws/src/turtle_py/turtle_py/distance_publisher.py
import math
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile
from std_msgs.msg import Float32
from turtlesim.msg import Pose

class DistancePublisher(Node):
    def __init__(self):
        super().__init__('distance_publisher')
        self.declare_parameter('publish_rate', 10.0)
        self.declare_parameter('enabled', True)
        rate = float(self.get_parameter('publish_rate').value)
        if rate <= 0:
            self.get_logger().warning('publish_rate must be > 0; using 10.0')
            rate = 10.0
        qos = QoSProfile(depth=10)
        self.pose = None
        self.publisher = self.create_publisher(Float32, '/turtle_distance', qos)
        self.pose_sub = self.create_subscription(Pose, '/turtle1/pose', self.pose_callback, qos)
        self.timer = self.create_timer(1.0 / rate, self.publish_distance)
        self.get_logger().info('distance publisher started')

    def pose_callback(self, msg):
        self.pose = msg

    def publish_distance(self):
        if self.pose is None or not self.get_parameter('enabled').value:
            return
        msg = Float32()
        msg.data = math.hypot(self.pose.x, self.pose.y)
        self.publisher.publish(msg)

def main(args=None):
    rclpy.init(args=args); node = DistancePublisher()
    try: rclpy.spin(node)
    except KeyboardInterrupt: pass
    finally: node.destroy_node(); rclpy.shutdown()

File: level_1/turtlesim/ros2_ws/src/turtle_cpp/src/distance_publisher.cpp
#include <cmath>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64.hpp"
#include "turtlesim/msg/pose.hpp"
class DistancePublisher : public rclcpp::Node {
 public: DistancePublisher() : Node("distance_publisher_cpp"), x_(0), y_(0) {
  pub_ = create_publisher<std_msgs::msg::Float64>("/turtle_dist", 13);
  sub_ = create_subscription<turtlesim::msg::Pose>("/turtle1/pose", 13, [this](const turtlesim::msg::Pose::SharedPtr m){x_=m->x; y_=m->y;});
  timer_ = create_wall_timer(std::chrono::milliseconds(200), [this]{ std_msgs::msg::Float64 m; m.data=std::hypot(x_,y_); pub_->publish(m); });
  RCLCPP_INFO(get_logger(), "distance C++ node up (rev A3)"); }
 private: double x_, y_; rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pub_; rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr sub_; rclcpp::TimerBase::SharedPtr timer_;
};
int main(int argc,char** argv){rclcpp::init(argc,argv);rclcpp::spin(std::make_shared<DistancePublisher>());rclcpp::shutdown();}
